### PR TITLE
More clean up and small optimisation

### PR DESCRIPTION
**Pull request by Dominaezzz**
### Pull Request Checklist

- [ ] Changes has been tested on an Android device or Android emulator with API 16
- [x] UI change has been tested on both light and dark themes
- [x] Pull request is based on the develop branch
- [ ] Pull request updates [CHANGES.md](https://github.com/vector-im/riotX-android/blob/develop/CHANGES.md)
- [x] Pull request includes screenshots or videos if containing UI changes
- [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#sign-off)

The usual Kotlin clean up.
I also left some comments pointing out race conditions.

About the 3rd commit, I saw this commit https://github.com/vector-im/riotX-android/commit/bcb811a7e83a56806a26e379f05bd8a5131085ff and thought to point out that a `Semaphore` can be used to limit the number of threads used for a transactions and will also reduce the number of `Dispatchers.Default` threads blocked for IO. Feel free to nuke the commit if it's not deemed useful.

### CODE REVIEW

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/api/Matrix.kt [hunk 1]
```diff
@@ -80,10 +80,12 @@ class Matrix private constructor(context: Context, matrixConfiguration: MatrixCo
                     val matrixConfiguration = (appContext as MatrixConfiguration.Provider).providesMatrixConfiguration()
                     instance = Matrix(appContext, matrixConfiguration)
                 } else {
+                    // TODO: Should set isInit back to false ?
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/im/vector/matrix/android/api/Matrix.kt:83]: Yes, probably, but when reaching that, the app will crash, the developer will fix, build and launch the app, so we may ignore

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/database/AsyncTransaction.kt [hunk 2]
```diff
@@ -19,27 +19,36 @@ import io.realm.Realm
...
-            if (isActive) {
-                bgRealm.commitTransaction()
-                val end = System.currentTimeMillis()
-                val time = end - start
-                Timber.v("Execute transaction in $time millis")
-            }
-        } finally {
-            if (bgRealm.isInTransaction) {
-                bgRealm.cancelTransaction()
+// TODO: Make this more of a global realm (or database) thing.
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/database/AsyncTransaction.kt:27]: I do not understand. @ganfra could you check the change to this file please? Thanks

Comment (Dominaezzz) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/database/AsyncTransaction.kt:27]: With this change, all calls to `awaitTransaction` will be synchronised, but if transactions are started without using `awaitTransaction` will run concurrently. So ideally everywhere should either always use `awaitTransaction` or use the `Semaphore`.

Although, if `awaitTransaction` is already consistently used everywhere for transactions, then that comment should be removed.

**Comment by bmarty** [comment]
Closed in favor of #2427, I think the other changes from this PR have already been implemented.

**Pull request closed without merging**